src/main/java/fi/aalto/cs/apluscourses/model/Module.java

```diff
@@ -13,32 +14,43 @@
   @NotNull
   private final String versionId;
 
+  @NotNull
+  private final ZonedDateTime downloadedAt;
```

Comment:
Shouldn't this be changed when the module is downloaded. Now it is only set when the object is created.

(This is my biggest concern about this PR.)

Reply:
yes, and initially I had an extra `update()` method in `Module`, but then realized it's better to let this open, so the person who'll implement the actual updating logics could decide. or what did you mean :question:

Reply:
OK, so now when started debugging I understand what you meant, sorry :student: . Clearly, the initially proposed approach with `Module.downloadAt` is incorrect. I'll probably try @nikke234 's proposed approach (analogous to `id` extraction).

:question: noticed one thing here... is it intended, that the `Module`s are being recreated every time project reloads? @OlliKiljunen @nikke234

Reply:
Well... yes and no.

It's intended in the sense that it is intentional and not a bug.

It's not intended in the sense that I'm not completely happy with the fact that they are recreated every time.

If someone can come up with a design that...
  + Doesn't recreate the module objects every time
  + Doesn't have any data races
  + Updates the tool window with new modules
  + Is relatively quick to implement and doesn't require massive refactoring

...then I would 100% support switching to that design. :smile: